[jit] add for in string support

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22006 [jit] add for in dict support
* **#21990 [jit] add for in string support**
* #21985 [jit] support iteration tuple unpacking
* #21950 [jit] refactor all for in range/tensor tests to be together with other for loop tests
* #21801 [jit] Tree based Iterator infrastructure: for in range/list/tensor/zip/enumerate

Differential Revision: [D15948547](https://our.internmc.facebook.com/intern/diff/D15948547)

wanchaol committed:
[jit] add for in string support

wanchaol committed:
Update on "[jit] add for in string support"

[jit] add for in string support

gh-metadata: pytorch pytorch 21990 gh/wanchaol/24/head

wanchaol committed:
rebase with master on "[jit] add for in string support"

[jit] add for in string support

gh-metadata: pytorch pytorch 21990 gh/wanchaol/24/head

zdevito commented on torch/csrc/jit/script/sugared_value.cpp:
```diff
@@ -280,6 +281,8 @@ Value* SimpleValue::getelem(const SourceRange&loc, Function& m, Value* i) {
   Value* cur_elem = nullptr;
   if (val_type->cast<ListType>()) {
     cur_elem = g.insert(aten::select, {val, i}, {}, loc);
+  } else if (val_type->cast<StringType>()) {
```
It would be nice to standardize a builtin like "__getelem__" to be used here rather than have a big if branch that emits different code, similar to our other magic methods. Furthermore, we should make serialization emit this as `foo[i]` so that we do not end up getting committed to preserving these operator names.

wanchaol replied:
I agree with this, will make a separate PR to define it. We will need to follow up on design the deprecation strategy of those prim operators that should be deleted.

wanchaol committed:
rebase with master on "[jit] add for in string support"

[jit] add for in string support

gh-metadata: pytorch pytorch 21990 gh/wanchaol/24/head

wanchaol committed:
rebase with master on "[jit] add for in string support"

[jit] add for in string support

gh-metadata: pytorch pytorch 21990 gh/wanchaol/24/head

wanchaol committed:
update comments on "[jit] add for in string support"

[jit] add for in string support

gh-metadata: pytorch pytorch 21990 gh/wanchaol/24/head

wanchaol committed:
Update on "[jit] add for in string support"

[jit] add for in string support

gh-metadata: pytorch pytorch 21990 gh/wanchaol/24/head

wanchaol committed:
Update on "[jit] add for in string support"

[jit] add for in string support

gh-metadata: pytorch pytorch 21990 gh/wanchaol/24/head
